### PR TITLE
Clarify operations on arrays

### PR DESCRIPTION
**Pull request by bcorrie**
As per the discussion in https://github.com/airr-community/airr-standards/issues/280#issuecomment-578860448

```
                "op":"OP",
                "content": {
                    "field":"study.keywords_study",
                    "value":"STRING"
                }
```
The semantics of this query are that for each element i in the array of N keywords_study the correct return value would be an OR operation on the results of N tests keywords_study[i] OP "STRING"

### CODE REVIEW

**Comment by bcorrie** [comment]
Please comment and/or add changes...

**Comment by bcorrie** [comment]
We need to determine if we are happy with these semantics when using the ```in``` and ```exclude``` operator.

study.keywords_study in ["contains_tcr", "contains_ig"] would return true if any of the array elements contain either value.

I don't think this works for operators that perform "negation" operations - not surprising since they are boolean operations I suppose 8-)

study.keywords_study exclude ["contains_tcr", "contains_ig"] would return true if any of the array elements exclude the values, which is probably not the intent...

Consider the array ["contain_single_cell", "contains_tcr"]

In the above scenario, we want to exclude this repertoire because it "contains_tcr" but the operation will perform an exclude of ["contains_tcr", "contains_ig"] on "contain_single_cell" and return true, and therefore the operation will return true...

**Comment by schristley** [comment]
The doc looks good to me. I'll add that `diagnosis`, `sample` and `data_processing` are also arrays.

Regarding `in` and `exclude`, is this how it actually works (that is you tested with queries) or how it might work?

**Comment by bcorrie** [comment]
> Regarding `in` and `exclude`, is this how it actually works (that is you tested with queries) or how it might work?

I was talking strictly semantics based on the definition above... I was more concerned about the logical expectation as to what should happen- haven't looked at any implementation yet...

The question came up when we were checking our implementation and it wasn't clear what should actually happen 8-)

**Comment by bcorrie** [comment]
> The doc looks good to me. I'll add that `diagnosis`, `sample` and `data_processing` are also arrays.

Is it expected that in the ADC API you would be able to perform a query on an object like this??? I have been assuming not... Hence it wasn't necessary to talk about these arrays from a query perspective... 

There is currently one other field array, data_processing_files - but as far as I am aware there are no other fields that one would be able to query on that are arrays (other than the possible discussion in #279)

**Comment by schristley** [comment]
> Is it expected that in the ADC API you would be able to perform a query on an object like this??? I have been assuming not... Hence it wasn't necessary to talk about these arrays from a query perspective...

Yes, like `sample.cell_subset` and all the other sample fields.

**Comment by schristley** [comment]
> study.keywords_study exclude ["contains_tcr", "contains_ig"] would return true if any of the array elements exclude the values, which is probably not the intent... I think the logical intent of this operation would be to exclude the data from this study if any of the keywords were present.

I see. Yes, the description you wrote is not really correct then. The `exclude` operator shouldn't work this way as an `OR`, it should work as the users would want it to, by [excluding documents](https://stackoverflow.com/questions/16221599/find-documents-with-arrays-not-containing-a-document-with-a-particular-field-val) with a specific value. However, we might be missing a test for this on an array. @bcorrie I'm not sure how you want to modify the docs, it might be better to put the array info along in the operator table instead of its own section, or maybe keep its own section but add table with each operator and its array behavior.

**Comment by bcorrie** [comment]
> > Is it expected that in the ADC API you would be able to perform a query on an object like this??? I have been assuming not... Hence it wasn't necessary to talk about these arrays from a query perspective...
> 
> Yes, like `sample.cell_subset` and all the other sample fields.

What I meant is performing a query on the field `sample` itself given that it is an array of objects.

That is, can you do something like:
```
                "op":"OP",
                "content": {
                    "field":"sample",
                    "value":SOMEVALUE
                }
```
Given that `sample` is an object, it is unclear what `SOMEVALUE` would be. This does not make much sense to me.

This is related to the NULL question that was raised in https://github.com/airr-community/airr-standards/issues/280#issuecomment-559452960 as that seems to be the only value that makes sense in this case. I could see *maybe* that you could do something like:

```
                "op":"OP",
                "content": {
                    "field":"sample",
                    "value":null
                }
```
That is essentially what you are talking about for subject.diagnosis in that you have set it to null in some cases, is that correct?

**Comment by schristley** [comment]
> What I meant is performing a query on the field sample itself given that it is an array of objects.

I see. No, I don't think many operators/queries would make sense directly on `sample`. The exception is maybe to check if it does or does not exist.

**Comment by bcorrie** [comment]
> > What I meant is performing a query on the field sample itself given that it is an array of objects.
> 
> I see. No, I don't think many operators/queries would make sense directly on `sample`. The exception is maybe to check if it does or does not exist.

I think this holds true of all Schema objects, correct, including `pcr_target` as @bussec was discussing? The bottom line here is that it is an error to try and apply an ADC API operator to a Schema object (AKA operators can only be applied to base types (not objects))

The one possible exception might be the "is missing" operator as per the discussion in #298 Does it make sense for us to allow for:
```
"op":"is missing", "content": {"field":"sample"}
```
or
```
"op":"is missing", "content": {"field":"sample.pcr_target"}
```
Personally, I think this is quite messy and potentially confusing... If a field is required but nullable, shouldn't it be present in the data as a null field? If you had a data set with 
```
{ ... "sample":[{ ... "pcr_target":null ... }] }
```
This would mean that the data was missing a required, nullable field (e.g. `sample.pcr_target.pcr_target_locus`) ... Doesn't that make it not AIRR compliant? I think it is an important part of the spec that these fields are required and should be provided (even as null) and we should avoid providing shortcuts to make it easy for researchers to leave things out...

**Comment by bcorrie** [comment]
Tests for array operations have been added.

https://github.com/airr-community/adc-api-tests/commit/da9e7503fffe1d4fb5ab9419e28ae82ba8cc4d97

**Comment by schristley** [comment]
> Tests for array operations have been added.

thanks, though I don't think that query works as you expect. When I run it on the test dataset in the reference implementation, it returns 0 results, which is what I would expect.

The `study.keywords_study` is on the whole study and not on individual repertoires. The study in the test dataset has both ig and tcr data so it's value is this:

```
"keywords_study":["contains_ig","contains_tcr"]
```

which would cause all those repertoires to be excluded.

**Comment by bcorrie** [comment]
> > Tests for array operations have been added.
> 
> thanks, though I don't think that query works as you expect. When I run it on the test dataset in the reference implementation, it returns 0 results, which is what I would expect.

Do you have a test server with the gold dataset that you are running this query against? I wrote the test with what I think is an expected result in my head. But I have nothing to test against as we don't have this implemented. If we exclude "contains_tcr" on a study with `study.keywords_study` ["contains_tcr","contains_ig"] (which is what Florian's data has) I would expect it to return 0 repertoires.

If that is what is returned then I think the query/filter is working. no? You are asking for all of the repertoires from studies that do not have TCR data in them... Since Florian's study has TCR data, then none of the repertoires are reported... so we get 0.
> though I don't think that query works as you expect.

Is the problem you are alluding to more about the fact that there are Repertoires in the Florian study that are not TCR - and therefore you would expect to have the IG repertoires reported from this query??? I can see how this might be confusing to the end user, but that expectation doesn't match the query that was issued. The query that would provide this information would be to ask for all of the Repertoires that exclude TCR data (rather than ask for all of the repertoires from the studies that exclude TCR data).

I don't think there is a direct mechanism to actually discover which repertoires are IG or TCR through the metadata that we have  - although we have had several discussions around that (#174 and #185) 8-) In fact that is where the `keywords_study` field came from in the first place... The cell_subset ontology is the closest thing we have at the moment...

Bottom line - if the query returns 0 on the Florian data, I think the query is correct...

**Comment by schristley** [comment]
> Do you have a test server with the gold dataset that you are running this query against?

Yes, but it is not publicly accessible. I just bring it up briefly to run tests then shut it down. I'm working on getting additional VMs so that I can have a more stable test environment, but that's going to take awhile. Even so, I still don't think I'm willing to open it to the public, the reference implementation has no security...

If you have a spare VM, getting up a reference implementation should be fairly straight forward (as in no harder than an iReceptor turnkey). I tried really hard to write complete and accurate [instructions](https://github.com/airr-community/adc-api#installation-procedure).

> If that is what is returned then I think the query/filter is working.

correct, I consider the query/filter to be working correctly.

> Is the problem you are alluding to ...

No, I'm just responding to your comment with the [test commit](https://github.com/airr-community/adc-api-tests/commit/da9e7503fffe1d4fb5ab9419e28ae82ba8cc4d97). I read it wrong and thought you were expecting it to return data. I agree with what you said, sounds like we are both on the same page.

I will say that having a test which returns 0 records isn't the best. There is always the chance that a bug causes all records to be excluded, which would be interpreted as a pass. Better for the test to split and return some data, though I'm not sure if there is another field we can test with, or we'd need to modify the dataset to create the condition.

**Comment by bcorrie** [comment]
> If you have a spare VM, getting up a reference implementation should be fairly straight forward (as in no harder than an iReceptor turnkey). I tried really hard to write complete and accurate [instructions](https://github.com/airr-community/adc-api#installation-procedure).

8-)

We are working on putting the Gold data set on an iReceptor Turnkey so we can use it to test our service API implementation... We should have that up soon... So we probably won't do that for the reference implementation... If we think that would be valuable in general, we could consider it...

**Comment by bcorrie** [comment]
> I will say that having a test which returns 0 records isn't the best. There is always the chance that a bug causes all records to be excluded, which would be interpreted as a pass. Better for the test to split and return some data, though I'm not sure if there is another field we can test with, or we'd need to modify the dataset to create the condition.

Agreed... I will try and figure out a query, but the problem is the Gold data set doesn't have a rich enough set of study_keyword. I suppose I can try and exclude of `['contains_ig', 'contains_paired_chain']`. This should return the set of repertoire that `'contains_tcr'` but not `'contains_ig'` with the `'contains_paired_chain'` being effectively a no-op as there aren't any of those to exclude...

**Comment by schristley** [comment]
Some repertoires have multiple entries for the `data_processing` array... And there are other enhancements we should add to the test data set as well, I went a created an [issue](https://github.com/airr-community/adc-api/issues/4) to be put it all together. Go ahead and add any other ideas you have.

**Comment by bcorrie** [comment]
@schristley can we also merge this, given that other discussions (tests for arrays - https://github.com/airr-community/adc-api/issues/4#issue-566518176), nullable schema objects - #328) are captured in other issues now.

**Comment by schristley** [comment]
@bcorrie yes except we didn't resolve how to document [exclude](https://github.com/airr-community/airr-standards/pull/322#issuecomment-581006136). I don't think `OR` is proper for `exclude`, it actually operates like an `AND`, that is, ALL entries in the array must exclude those values for the document to be returned.

Works a bit like deMorgan's law:

`in`: value OR value OR value
`exclude`: !value AND !value AND !value

It's also similar for the `=` and `!=` operators.

**Comment by bcorrie** [comment]
@schristley OK to delete this branch?
